applications/FluidDynamicsApplication/python_scripts/fluid_dynamics_analysis.py

```diff
@@ -0,0 +1,240 @@
+        else:
+            self.solver.AddVariables()
+            self.solver.ImportModelPart()
+            self.solver.AddDofs()
+
+        self.model_part = self.solver.GetComputingModelPart()
+
+        # Fill a Model instance using input
+        self.model = Model()
+        self.model.AddModelPart(self.main_model_part)
+
+    def SetUpConditions(self):
+        '''Read the boundary and initial conditions for the problem and initialize the processes that will manage them.'''
+
+        ## Processes construction
+        from process_factory import KratosProcessFactory
+        factory = KratosProcessFactory(self.model)
+        # The list of processes will contain a list with each individual process already constructed (boundary conditions, initial conditions and gravity)
+        # Note 1: gravity is constructed first. Outlet process might need its information.
+        # Note 2: initial conditions are constructed before BCs. Otherwise, they may overwrite the BCs information.
+        self.simulation_processes =  factory.ConstructListOfProcesses( self.project_parameters["gravity"] )
+        self.simulation_processes += factory.ConstructListOfProcesses( self.project_parameters["initial_conditions_process_list"] )
+        self.simulation_processes += factory.ConstructListOfProcesses( self.project_parameters["boundary_conditions_process_list"] )
+        self.simulation_processes += factory.ConstructListOfProcesses( self.project_parameters["auxiliar_process_list"] )
+
+        for process in self.simulation_processes:
+            process.ExecuteInitialize()
+
+    def SetUpAnalysis(self):
+        '''
+        Initialize the Python solver and its auxiliary tools and processes.
+        This function should leave everything prepared so that the simulation
+        can start immediately after exiting it.
+        '''
+
+        self.solver.Initialize()
+
+        #TODO this should be generic
+        # initialize GiD  I/O
+        self._SetUpGiDOutput()
+
+        ## Writing the full ProjectParameters file before solving
+        if self.is_printing_rank and self.echo_level > 1:
+            with open("ProjectParametersOutput.json", 'w') as parameter_output_file:
+                parameter_output_file.write(self.project_parameters.PrettyPrintJsonString())
+
+        ## Stepping and time settings
+        self.end_time = self.project_parameters["problem_data"]["end_time"].GetDouble()
+
+        if self.main_model_part.ProcessInfo[IS_RESTARTED]:
+            self.time = self.main_model_part.ProcessInfo[TIME]
+            self.step = self.main_model_part.ProcessInfo[STEP]
+        else:
+            self.time = 0.0
+            self.step = 0
+
+        for process in self.simulation_processes:
+            process.ExecuteBeforeSolutionLoop()
+            
+        if self.have_output:
+            self.output.ExecuteBeforeSolutionLoop()
+
+
+    def _SetUpGiDOutput(self):
+        '''Initialize self.output as a GiD output instance.'''
```

Comment:
I think I get it now. There are the extra customization points ExecuteAfterOutputStep, etc.

Reply:
At the moment, the GiDOutput is not a "real" process, since it has one extra method (`PrintOutput`) not included in the standard process public interface. I agree that this could be used to make the class compact/readable, though.